### PR TITLE
Update base image to stretch and OpenSSL to 1.1.0

### PR DESCRIPTION
**Pull request by J0WI**

### CODE REVIEW

File: 2.4/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM debian:jessie-backports
+FROM debian:stretch-backports
```
Comment (tianon) [inline 2.4/Dockerfile]: Updating to `stretch`, we shouldn't need `-backports` anymore:

```suggestion
FROM debian:stretch-slim
```

File: 2.4/Dockerfile [hunk 2]
```diff
@@ -11,22 +11,7 @@ WORKDIR $HTTPD_PREFIX
...
-		echo 'Pin: release n=stretch'; \
-		echo 'Pin-Priority: -10'; \
-		echo; \
-# except nghttp2, which is the reason we're here
-		echo 'Package: libnghttp2*'; \
-		echo "Pin: version $NGHTTP2_VERSION"; \
-		echo 'Pin-Priority: 990'; \
-		echo; \
-	} > /etc/apt/preferences.d/unstable-nghttp2
+ENV OPENSSL_VERSION 1.1.0f-3+deb9u2
```
Comment (tianon) [inline 2.4/Dockerfile]: I think both these environment variables can and ought to be dropped now -- we added them to ensure we got the appropriate versions from `stretch` and/or `jessie-backports`, but that's not really necessary if we're already building from the stable `stretch` base.

File: 2.4/Dockerfile [hunk 3]
```diff
@@ -39,8 +24,8 @@ RUN apt-get update \
 		libaprutil1-dev \
 		liblua5.2-0 \
 		libnghttp2-14=$NGHTTP2_VERSION \
-		libpcre++0 \
-		libssl1.0.0=$OPENSSL_VERSION \
+		libpcre++0v5 \
+		libssl1.1=$OPENSSL_VERSION \
 		libxml2 \
```
Comment (tianon) [inline 2.4/Dockerfile]: It's not something we should do for _this_ PR, but I think in the future we probably ought to consider something like https://github.com/docker-library/php/blob/32575d71484b97b6a870527953c986a4c076b99b/Dockerfile-debian.template#L185-L196 so we don't have to hand-maintain this list (which is admittedly fragile and error-prone).

File: update.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ fi
 versions=( "${versions[@]%/}" )
 
 nghttp2VersionDebian="$(docker run -i --rm debian:stretch-slim bash -c 'apt-get update -qq && apt-cache show "$@"' -- 'libnghttp2-dev' |tac|tac| awk -F ': ' '$1 == "Version" { print $2; exit }')"
-opensslVersionDebian="$(docker run -i --rm debian:jessie-backports bash -c 'apt-get update -qq && apt-cache show "$@"' -- 'openssl' |tac|tac| awk -F ': ' '$1 == "Version" { print $2; exit }')"
+opensslVersionDebian="$(docker run -i --rm debian:stretch-backports bash -c 'apt-get update -qq && apt-cache show "$@"' -- 'openssl' |tac|tac| awk -F ': ' '$1 == "Version" { print $2; exit }')"
```
Comment (tianon) [inline update.sh]: This should drop `-backports` too, but we should drop both these entirely instead.

I'm happy to take on that change if you'd prefer. :+1:

**Comment by J0WI** [comment]
Shouldn't we keep `-backports` to be ready for OpenSSL 1.1.1 (TLS 1.3 support)?

**Comment by tianon** [comment]
Looking at https://tracker.debian.org/pkg/openssl, I remain wary of our ability to use `openssl` version 1.1.1 in Debian Stretch (it appears to be known to break a non-trivial number of packages; see https://bugs.debian.org/907015).

**Comment by J0WI** [comment]
> The Ubuntu project is currently evaluating replacing OpenSSL 1.1.0 with backported OpenSSL 1.1.1 in Ubuntu 18.04 LTS.

https://bugs.launchpad.net/ubuntu/+source/openssl/+bug/1797386

So Debian may be able to backport it too. However, I changed the base to `-slim` for now.

**Comment by tianon** [comment]
With this latest push, size difference is ~178MB down to ~132MB.

If I switch from `stretch-slim` to simply `stretch` (to have a more fair size comparison test), it's ~190MB (which is the jessie->stretch conversion, so not totally unexpected).

**Pull request merged**